File: app/src/main/java/com/mikepenz/fastadapter/app/items/SwipeableItem.java
package com.mikepenz.fastadapter.app.items;

import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.TextView;

import com.mikepenz.fastadapter_extensions.drag.IDraggable;
import com.mikepenz.fastadapter.IItem;
import com.mikepenz.fastadapter_extensions.swipe.ISwipeable;
import com.mikepenz.fastadapter.app.R;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.materialdrawer.holder.StringHolder;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Mattias on 2016-02-15.
 */
public class SwipeableItem extends AbstractItem<SwipeableItem, SwipeableItem.ViewHolder> implements ISwipeable<SwipeableItem, IItem>, IDraggable<SwipeableItem, IItem> {

    public StringHolder name;
    public StringHolder description;

    public StringHolder undoTextSwipeFromRight;
    public StringHolder undoTextSwipeFromLeft;
    public StringHolder undoTextSwipeFromTop;
    public StringHolder undoTextSwipeFromBottom;

    public int swipedDirection;
    private Runnable swipedAction;
    public boolean swipeable = true;
    public boolean draggable = true;

    public SwipeableItem withName(String Name) {
        this.name = new StringHolder(Name);
        return this;
    }

    public SwipeableItem withName(@StringRes int NameRes) {
        this.name = new StringHolder(NameRes);
        return this;
    }

    public SwipeableItem withDescription(String description) {
        this.description = new StringHolder(description);
        return this;
    }

    public SwipeableItem withDescription(@StringRes int descriptionRes) {
        this.description = new StringHolder(descriptionRes);
        return this;
    }

    @Override
    public boolean isSwipeable() {
        return swipeable;
    }

    @Override
    public SwipeableItem withIsSwipeable(boolean swipeable) {
        this.swipeable = swipeable;
        return this;
    }

    @Override
    public boolean isDraggable() {
        return draggable;
    }

    @Override
    public SwipeableItem withIsDraggable(boolean draggable) {
        this.draggable = draggable;
        return this;
    }

    public void setSwipedDirection(int swipedDirection) {
        this.swipedDirection = swipedDirection;
    }

    public void setSwipedAction(Runnable action) {
        this.swipedAction = action;
    }

    /**
     * defines the type defining this item. must be unique. preferably an id
     *
     * @return the type
     */
    @Override
    public int getType() {
        return R.id.fastadapter_swipable_item_id;
    }

    /**
     * defines the layout which will be used for this item in the list
     *
     * @return the layout for this item
     */
    @Override
    public int getLayoutRes() {
        return R.layout.swipeable_item;
    }

    /**
     * binds the data of this item onto the viewHolder
     *
     * @param viewHolder the viewHolder of this item
     */
    @Override
    public void bindView(ViewHolder viewHolder, List<Object> payloads) {
        super.bindView(viewHolder, payloads);

        //set the text for the name
        StringHolder.applyTo(name, viewHolder.name);
        //set the text for the description or hide
        StringHolder.applyToOrHide(description, viewHolder.description);

        viewHolder.swipeResultContent.setVisibility(swipedDirection != 0 ? View.VISIBLE : View.GONE);
        viewHolder.itemContent.setVisibility(swipedDirection != 0 ? View.GONE : View.VISIBLE);

        CharSequence swipedAction = null;
        CharSequence swipedText = null;
        if (swipedDirection != 0) {
            swipedAction = viewHolder.itemView.getContext().getString(R.string.action_undo);
            swipedText = swipedDirection == ItemTouchHelper.LEFT ? "Removed" : "Archived";
            viewHolder.swipeResultContent.setBackgroundColor(ContextCompat.getColor(viewHolder.itemView.getContext(), swipedDirection == ItemTouchHelper.LEFT ? R.color.md_red_900 : R.color.md_blue_900));
        }
        viewHolder.swipedAction.setText(swipedAction == null ? "" : swipedAction);
        viewHolder.swipedText.setText(swipedText == null ? "" : swipedText);
        viewHolder.swipedActionRunnable = this.swipedAction;
    }

    @Override
    public void unbindView(ViewHolder holder) {
        super.unbindView(holder);
        holder.name.setText(null);
        holder.description.setText(null);
        holder.swipedAction.setText(null);
        holder.swipedText.setText(null);
        holder.swipedActionRunnable = null;
    }

    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    /**
     * our ViewHolder
     */
    protected static class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.material_drawer_name)
        TextView name;
        @BindView(R.id.material_drawer_description)
        TextView description;
        @BindView(R.id.swipe_result_content)
        View swipeResultContent;
        @BindView(R.id.item_content)
        View itemContent;
        @BindView(R.id.swiped_text)
        TextView swipedText;
        @BindView(R.id.swiped_action)
        TextView swipedAction;

        public Runnable swipedActionRunnable;

        public ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
            swipedAction.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (swipedActionRunnable != null) {
                        swipedActionRunnable.run();
                    }
                }
            });
        }
    }
}

File: app/src/main/java/com/mikepenz/fastadapter/app/items/SwipeableItem.kt
package com.mikepenz.fastadapter.app.items

import android.graphics.Color
import android.view.View
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.app.R
import com.mikepenz.fastadapter.app.adapters.IDraggableViewHolder
import com.mikepenz.fastadapter.drag.IDraggable
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter.swipe.ISwipeable
import com.mikepenz.materialdrawer.holder.StringHolder

/**
 * Created by Mattias on 2016-02-15.
 */
class SwipeableItem : AbstractItem<SwipeableItem.ViewHolder>(), ISwipeable, IDraggable {

    var name: StringHolder? = null
    var description: StringHolder? = null

    var undoTextSwipeFromRight: StringHolder? = null
    var undoTextSwipeFromLeft: StringHolder? = null
    var undoTextSwipeFromTop: StringHolder? = null
    var undoTextSwipeFromBottom: StringHolder? = null

    var swipedDirection: Int = 0
    var swipedAction: Runnable? = null
    override var isSwipeable = true
    override var isDraggable = true

    /**
     * defines the type defining this item. must be unique. preferably an id
     *
     * @return the type
     */
    override val type: Int
        get() = R.id.fastadapter_swipable_item_id

    /**
     * defines the layout which will be used for this item in the list
     *
     * @return the layout for this item
     */
    override val layoutRes: Int
        get() = R.layout.swipeable_item

    fun withName(Name: String): SwipeableItem {
        this.name = StringHolder(Name)
        return this
    }

    fun withName(@StringRes NameRes: Int): SwipeableItem {
        this.name = StringHolder(NameRes)
        return this
    }

    fun withDescription(description: String): SwipeableItem {
        this.description = StringHolder(description)
        return this
    }

    fun withDescription(@StringRes descriptionRes: Int): SwipeableItem {
        this.description = StringHolder(descriptionRes)
        return this
    }

    fun withIsSwipeable(swipeable: Boolean): SwipeableItem {
        this.isSwipeable = swipeable
        return this
    }

    fun withIsDraggable(draggable: Boolean): SwipeableItem {
        this.isDraggable = draggable
        return this
    }

    /**
     * binds the data of this item onto the viewHolder
     *
     * @param holder the viewHolder of this item
     */
    override fun bindView(holder: ViewHolder, payloads: List<Any>) {
        super.bindView(holder, payloads)

        //set the text for the name
        StringHolder.applyTo(name, holder.name)
        //set the text for the description or hide
        StringHolder.applyToOrHide(description, holder.description)

        holder.swipeResultContent.visibility = if (swipedDirection != 0) View.VISIBLE else View.GONE
        holder.itemContent.visibility = if (swipedDirection != 0) View.GONE else View.VISIBLE

        var swipedAction: CharSequence? = null
        var swipedText: CharSequence? = null
        if (swipedDirection != 0) {
            swipedAction = holder.itemView.context.getString(R.string.action_undo)
            swipedText = if (swipedDirection == ItemTouchHelper.LEFT) "Removed" else "Archived"
            holder.swipeResultContent.setBackgroundColor(if (swipedDirection == ItemTouchHelper.LEFT) Color.RED else Color.BLUE)
        }
        holder.swipedAction.text = swipedAction ?: ""
        holder.swipedText.text = swipedText ?: ""
        holder.swipedActionRunnable = this.swipedAction
    }

    override fun unbindView(holder: ViewHolder) {
        super.unbindView(holder)
        holder.name.text = null
        holder.description.text = null
        holder.swipedAction.text = null
        holder.swipedText.text = null
        holder.swipedActionRunnable = null
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    /**
     * our ViewHolder
     */
    class ViewHolder(view: View) : RecyclerView.ViewHolder(view), IDraggableViewHolder {
        var name: TextView = view.findViewById(R.id.material_drawer_name)
        var description: TextView = view.findViewById(R.id.material_drawer_description)
        var swipeResultContent: View = view.findViewById(R.id.swipe_result_content)
        var itemContent: View = view.findViewById(R.id.item_content)
        var swipedText: TextView = view.findViewById(R.id.swiped_text)
        var swipedAction: TextView = view.findViewById(R.id.swiped_action)

        var swipedActionRunnable: Runnable? = null

        init {
            swipedAction.setOnClickListener {
                swipedActionRunnable?.run()
            }
        }

        override fun onDropped() {
            itemContent.setBackgroundColor(Color.TRANSPARENT)
        }

        override fun onDragged() {
            itemContent.setBackgroundColor(Color.LTGRAY)
        }
    }
}
